### PR TITLE
Fix that unsigned with zerofill is treated as signed

### LINKED ISSUE #27125: MySQL unsigned column is treated as signed column  when column is like 'unsigned zerofill'
### Steps to reproduce
 i'm using rails without migration.
The problem occurs in table like:
``` sql
# table definition
CREATE TABLE `samples` (
  `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
  `problem_column` mediumint(7) unsigned zerofill DEFAULT NULL,
   PRIMARY KEY (`id`)
) ENGINE=InnoDB  DEFAULT CHARSET=utf8;
```

problem_column is unsigned, but Rails treat it as signed

``` ruby
# model
class Sample < ApplicationRecord
end
```
### Expected behavior
recognize unsigned column as so not only case it is at the end of sentence.

### Actual behavior
recognize unsigned column as so "only" case it is at the end of sentence(because of https://github.com/rails/rails/blob/master/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb#L694 and https://github.com/rails/rails/blob/master/activerecord/lib/active_record/connection_adapters/mysql/column.rb#L7).

``` ruby
Sample.columns[-1]
#=> #<ActiveRecord::ConnectionAdapters::MySQL::Column:0x007fd91a3a9328 @name="problem_column", @table_name="samples", @sql_type_metadata=#<ActiveRecord::ConnectionAdapters::SqlTypeMetadata:0x007fd91a3a9580 @sql_type="mediumint(7) unsigned zerofill", @type=:integer, @limit=3, @precision=nil, @scale=nil>, @null=true, @default=nil, @default_function=nil, @collation=nil, @comment=nil>

Sample.columns[-1].unsigned?
#=> false

Sample.create(problem_column: 16777215)
# ActiveModel::RangeError: 16777215 is out of range for ActiveModel::Type::Integer with limit 3
```

### System configuration
**Rails version**: 5.0

**Ruby version**: 2.3.0

### PR DESCRIPTION
**Pull request by kamipo**
Fixes #27125.

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/mysql/column.rb [hunk 1]
```diff
@@ -5,7 +5,7 @@ class Column < ConnectionAdapters::Column # :nodoc:
         delegate :extra, to: :sql_type_metadata, allow_nil: true
 
         def unsigned?
-          /\bunsigned\z/.match?(sql_type)
+          /\bunsigned\b/.match?(sql_type)
```
Comment (prathamesh-sonpatki) [inline activerecord/lib/active_record/connection_adapters/mysql/column.rb]: @kamipo The enum unsigned test will fail due to this, I think we need special treatment for enum.

Something like this:

```ruby
        def unsigned?
          if (/^enum/).match?(sql_type)
            actual_sql_type = sql_type[/^enum\((.+)\)(.+)/i, 2]
          else
            actual_sql_type = sql_type
          end

          /\bunsigned\b/.match?(actual_sql_type)
        end
```

**Review by utilum — COMMENTED** [review]
Would it be nice to give this all tests that other types need to pass?

File: activerecord/lib/active_record/connection_adapters/mysql/column.rb [hunk 2]
```diff
@@ -5,7 +5,7 @@ class Column < ConnectionAdapters::Column # :nodoc:
         delegate :extra, to: :sql_type_metadata, allow_nil: true
 
         def unsigned?
-          /\bunsigned\z/.match?(sql_type)
+          !/\A(?:enum|set)\b/.match?(sql_type) && /\bunsigned\b/.match?(sql_type)
```
Comment (prathamesh-sonpatki) [inline activerecord/lib/active_record/connection_adapters/mysql/column.rb]: Nice!

Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/mysql/column.rb]: I don't get why this is needed. Does `enum` and `set` types allow to be defined as `unsigned`?

Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/mysql/column.rb]: `enum` and `set` types does not allow to be defined as `unsigned`. But if defined as `enum('foo','bar','unsigned')`, `column.unsigned?` returns `true`. It is incorrect.

https://github.com/rails/rails/blob/master/activerecord/test/cases/adapters/mysql2/enum_test.rb#L12-L15

```ruby
  def test_should_not_be_unsigned
    column = EnumTest.columns_hash["enum_column"]
    assert_not column.unsigned?
  end
```

**Review by rafaelfranca — CHANGES_REQUESTED** [review]
Could you add a CHANGELOG entry?

**Comment by kamipo** [comment]
I added a CHANGELOG entry!

**Comment by rafaelfranca** [comment]
Backported in f39779f

**Pull request merged**